### ISSUE TITLE
How to store test_step outputs to file?

### ISSUE THREAD

**Issue by Ceceu** [PyTorchLightning/pytorch-lightning#3705]
Is there an approach to save to one file all the outputs during `test_step`?

```python
    def test_step(self, batch, batch_idx):
        x1, x2 = batch["x1"], batch["x2"]
        r1, r2 = self(x1, x2)
        
        test_loss = self.loss_fn(predict)
        test_mrr = self.mrr(r1, r2)

        return {'test_loss': test_loss, 'test_mrr': test_mrr}
```
My `LightningModule` outputs two dense vectors for representation `r1` and `r2`. They must be saved to be used in a downstream task and for the measurement of some metrics.

**Comment by rohitgr7** [comment]
[EvalResult.write_dict()](https://pytorch-lightning.readthedocs.io/en/latest/results.html#evalresult-api)

**Comment by Ceceu** [comment]
> [EvalResult.write_dict()](https://pytorch-lightning.readthedocs.io/en/latest/results.html#evalresult-api)

I am sorry but I could not understand how to use `TrainResult` or `EvalResult` to save on file the outputs of the `LightningModule`.

**Comment by rohitgr7** [comment]
```python
def test_step(self, batch, batch_idx):
        x1, x2 = batch["x1"], batch["x2"]
        r1, r2 = self(x1, x2)

        # saving prediction to file?
        predictions_to_write = {'r1': r1, 'r2': r2}
        result.write_dict(predictions_to_write, filename)
        
        # log or attach other metrics
        result.log('test_loss', test_loss)
        # or
        result.test_loss = test_loss
        return result
```

**Comment by ydcjeff** [comment]
Closing it as it is resolved, feel free to reopen if needed

**Comment by Ceceu** [comment]
Hello @rohitgr7,

As `EvalResult` and `TrainResult` is not supported anymore (after [pr 3938](https://github.com/PyTorchLightning/pytorch-lightning/pull/3968)), how could someone save predictions to file?

**Comment by rohitgr7** [comment]
https://github.com/PyTorchLightning/pytorch-lightning/issues/3705#issuecomment-700313649

This should work in validation and testing but for training its not supported yet.

**Comment by Ceceu** [comment]
Thank you @rohitgr7,
Then, could I use `test_step` to save `predictions_dict`, reload it from disk in `test_epoch_end`, and use reloaded `predictions_dict` to calculate a metric?

**Comment by rohitgr7** [comment]
@Ceceu I think it aggregates everything and saves it after epoch_end, but not sure if I am correct here. Maybe try and tell??
